File: client/lib/models/troop.dart
import 'package:json_annotation/json_annotation.dart';
import '../game/hex/cube_coord.dart';
import 'enums.dart';

part 'troop.g.dart';

@JsonSerializable()
class Troop {
  final String id;
  final TroopType type;
  @JsonKey(name: 'owner_id')
  final String ownerId;
  final CubeCoord hex;
  @JsonKey(name: 'current_hp')
  final int currentHp;
  @JsonKey(name: 'max_hp')
  final int maxHp;
  final int atk;
  final int def;
  final int mobility;
  final int range;
  final String damage;
  @JsonKey(name: 'is_ready')
  final bool isReady;
  @JsonKey(name: 'has_moved')
  final bool hasMoved;
  @JsonKey(name: 'has_attacked')
  final bool hasAttacked;
  @JsonKey(name: 'was_in_combat')
  final bool wasInCombat;
  @JsonKey(name: 'remaining_mobility')
  final int remainingMobility;

  const Troop({
    required this.id,
    required this.type,
    required this.ownerId,
    required this.hex,
    required this.currentHp,
    required this.maxHp,
    required this.atk,
    required this.def,
    required this.mobility,
    required this.range,
    required this.damage,
    required this.isReady,
    required this.hasMoved,
    required this.hasAttacked,
    required this.wasInCombat,
    required this.remainingMobility,
  });

  bool get isAlive => currentHp > 0;
  bool get canAct => isAlive && isReady;
  bool get canMove => canAct && !hasMoved && remainingMobility > 0;
  bool get canAttack => canAct && !hasAttacked;
  bool get isMelee => range == 1;

  factory Troop.fromJson(Map<String, dynamic> json) => _$TroopFromJson(json);
  Map<String, dynamic> toJson() => _$TroopToJson(this);

  Troop copyWith({
    String? id,
    TroopType? type,
    String? ownerId,
    CubeCoord? hex,
    int? currentHp,
    int? maxHp,
    int? atk,
    int? def,
    int? mobility,
    int? range,
    String? damage,
    bool? isReady,
    bool? hasMoved,
    bool? hasAttacked,
    bool? wasInCombat,
    int? remainingMobility,
  }) {
    return Troop(
      id: id ?? this.id,
      type: type ?? this.type,
      ownerId: ownerId ?? this.ownerId,
      hex: hex ?? this.hex,
      currentHp: currentHp ?? this.currentHp,
      maxHp: maxHp ?? this.maxHp,
      atk: atk ?? this.atk,
      def: def ?? this.def,
      mobility: mobility ?? this.mobility,
      range: range ?? this.range,
      damage: damage ?? this.damage,
      isReady: isReady ?? this.isReady,
      hasMoved: hasMoved ?? this.hasMoved,
      hasAttacked: hasAttacked ?? this.hasAttacked,
      wasInCombat: wasInCombat ?? this.wasInCombat,
      remainingMobility: remainingMobility ?? this.remainingMobility,
    );
  }
}

File: server/internal/model/troop.go
package model

import (
	"github.com/teomiscia/hexbattle/internal/hex"
)

// Troop represents a single unit on the battlefield.
type Troop struct {
	ID                string    `json:"id"`
	Type              TroopType `json:"type"`
	OwnerID           string    `json:"owner_id"`
	Hex               hex.Coord `json:"hex"`
	CurrentHP         int       `json:"current_hp"`
	MaxHP             int       `json:"max_hp"`
	ATK               int       `json:"atk"`
	DEF               int       `json:"def"`
	Mobility          int       `json:"mobility"`
	Range             int       `json:"range"`
	Damage            string    `json:"damage"` // Dice notation, e.g. "1D6+1"
	IsReady           bool      `json:"is_ready"`
	HasMoved          bool      `json:"has_moved"`
	HasAttacked       bool      `json:"has_attacked"`
	WasInCombat       bool      `json:"was_in_combat"`
	RemainingMobility int       `json:"remaining_mobility"`
}

// IsAlive returns true if the troop has HP remaining.
func (t *Troop) IsAlive() bool {
	return t.CurrentHP > 0
}

// CanAct returns true if the troop is ready and alive.
func (t *Troop) CanAct() bool {
	return t.IsAlive() && t.IsReady
}

// CanMove returns true if the troop can still move this turn.
func (t *Troop) CanMove() bool {
	return t.CanAct() && !t.HasMoved && t.RemainingMobility > 0
}

// CanAttack returns true if the troop can still attack this turn.
func (t *Troop) CanAttack() bool {
	return t.CanAct() && !t.HasAttacked
}

// ResetForTurn resets per-turn action flags and restores mobility.
func (t *Troop) ResetForTurn() {
	t.HasMoved = false
	t.HasAttacked = false
	t.WasInCombat = false
	t.RemainingMobility = t.Mobility
}

// TakeDamage applies damage to the troop and returns true if it died.
func (t *Troop) TakeDamage(amount int) bool {
	t.CurrentHP -= amount
	if t.CurrentHP < 0 {
		t.CurrentHP = 0
	}
	return t.CurrentHP <= 0
}

// Heal restores HP up to MaxHP. Returns the amount actually healed.
func (t *Troop) Heal(amount int) int {
	before := t.CurrentHP
	t.CurrentHP += amount
	if t.CurrentHP > t.MaxHP {
		t.CurrentHP = t.MaxHP
	}
	return t.CurrentHP - before
}

// IsMelee returns true if the troop has a range of 1.
func (t *Troop) IsMelee() bool {
	return t.Range == 1
}
